pkg/kubelet/kubelet.go

```diff
@@ -759,6 +765,61 @@ func (kl *Kubelet) makeEnvironmentVariables(ns string, container *api.Container)
 	return result, nil
 }
 
+// runtimeEnvVarValue determines the value that an env var should take when a container
+// is started.  If the value of the env var is the empty string, the source of the env var
+// is resolved, if one is specified.
+//
+// TODO: preliminary factoring; make better
+func (kl *Kubelet) runtimeEnvVarValue(envVar api.EnvVar, pod *api.Pod) (string, error) {
+	runtimeVal := envVar.Value
+	if runtimeVal != "" {
+		return runtimeVal, nil
+	}
+
+	switch {
+	case envVar.ObjectInfo != nil:
+		return kl.objectInfoRuntimeValue(envVar.ObjectInfo, pod)
+	case envVar.HostInfo != nil:
+		return kl.hostInfoRuntimeValue(envVar.HostInfo)
+	}
+
+	return runtimeVal, nil
+}
+
+func (kl *Kubelet) objectInfoRuntimeValue(objInfo *api.ObjectReference, pod *api.Pod) (string, error) {
+	// TODO: use objInfo Kind instead of "Pod"
+	internalFieldPath, _, err := api.Scheme.ConvertFieldLabel(objInfo.APIVersion, "Pod", objInfo.FieldPath, "")
+	if err != nil {
+		return "", err
+	}
+
+	runtimeVal := ""
+
+	// TODO: currently there is no generic utility for retrieving a field of
+	// an object using a fieldPath expression.  For now, support only the
+	// Name and Namespace fields.
+	switch internalFieldPath {
+	case "metadata.name":
```

Comment:
This reminds me, there should really be a spec for what we come up with
that contains enough information to reimplement. The blog posts for
jsonpath aren't enough IMO. They're good blog posts but not a spec.
On Thu, Apr 16, 2015 at 12:05 AM Brian Grant <notifications@github.com>
wrote:

> In pkg/kubelet/kubelet.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6739#discussion_r28482739>
> :
>
> > +}
> > +
> > +func (kl *Kubelet) objectInfoRuntimeValue(objInfo *api.ObjectReference, pod *api.Pod) (string, error) {
> > +	// TODO: use objInfo Kind instead of "Pod"
> > +	internalFieldPath, _, err := api.Scheme.ConvertFieldLabel(objInfo.APIVersion, "Pod", objInfo.FieldPath, "")
> > +	if err != nil {
> > +		return "", err
> > +	}
> > +
> > +	runtimeVal := ""
> > +
> > +	// TODO: currently there is no generic utility for retrieving a field of
> > +	// an object using a fieldPath expression.  For now, support only the
> > +	// Name and Namespace fields.
> > +	switch internalFieldPath {
> > +	case "metadata.name":
>
> I know we're currently using this syntax for other field specifications,
> but I'd prefer to go with something jsonpath-compatible. My reading of
> http://goessner.net/articles/JsonPath/ is that it would be: $.
> metadata.name.
>
> If we're going to break compatibility, we should do it sooner rather than
> later.
>
> @lavalamp <https://github.com/lavalamp> Any objections to changing field
> specifications to jsonpath syntax? Annoyingly, it's not the same as jq
> syntax, and there seem to be several other variants. However, all of the
> variants I've seen at least start paths with a .. Jsonpath and a few
> others start with $.. AFAICT, jsonpath is the most commonly mentioned.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6739/files#r28482739>
> .
>